server/src/main/java/org/elasticsearch/index/engine/TranslogOperationAsserter.java

```diff
@@ -0,0 +1,127 @@
+                    return super.sameIndexOperation(synthesizeSource(engineConfig, o1), o2)
+                        || super.sameIndexOperation(o1, synthesizeSource(engineConfig, o2));
+                }
+                return false;
+            }
+        };
+    }
+
+    static Translog.Index synthesizeSource(EngineConfig engineConfig, Translog.Index op) throws IOException {
+        final ShardId shardId = engineConfig.getShardId();
+        MappingLookup mappingLookup = engineConfig.getMapperService().mappingLookup();
+        DocumentParser documentParser = engineConfig.getMapperService().documentParser();
+        try (var reader = new TranslogDirectoryReader(shardId, op, mappingLookup, documentParser, engineConfig, () -> {})) {
+            final Engine.Searcher searcher = new Engine.Searcher(
+                "assert_translog",
+                reader,
+                engineConfig.getSimilarity(),
+                engineConfig.getQueryCache(),
+                engineConfig.getQueryCachingPolicy(),
+                () -> {}
+            );
+            try (
+                LuceneSyntheticSourceChangesSnapshot snapshot = new LuceneSyntheticSourceChangesSnapshot(
+                    mappingLookup,
+                    searcher,
+                    LuceneSyntheticSourceChangesSnapshot.DEFAULT_BATCH_SIZE,
+                    Integer.MAX_VALUE,
+                    op.seqNo(),
+                    op.seqNo(),
+                    true,
+                    false,
+                    engineConfig.getIndexSettings().getIndexVersionCreated()
+                )
+            ) {
+                final Translog.Operation normalized = snapshot.next();
+                assert normalized != null : "expected one operation; got zero";
+                return (Translog.Index) normalized;
+            }
+        }
+    }
+
+    boolean sameIndexOperation(Translog.Index o1, Translog.Index o2) throws IOException {
+        // TODO: We haven't had timestamp for Index operations in Lucene yet, we need to loosen this check without timestamp.
+        return Objects.equals(o1.id(), o2.id())
+            && Objects.equals(o1.source(), o2.source())
+            && Objects.equals(o1.routing(), o2.routing())
+            && o1.primaryTerm() == o2.primaryTerm()
+            && o1.seqNo() == o2.seqNo()
+            && o1.version() == o2.version();
+    }
+
+    public boolean assertSameOperations(
+        long seqNo,
+        long generation,
+        Translog.Operation newOp,
+        Translog.Operation prvOp,
+        Exception prevFailure
+    ) throws IOException {
+        final boolean sameOp;
+        if (newOp instanceof final Translog.Index o2 && prvOp instanceof final Translog.Index o1) {
+            sameOp = sameIndexOperation(o1, o2);
+        } else if (newOp instanceof final Translog.Delete o1 && prvOp instanceof final Translog.Delete o2) {
+            sameOp = Objects.equals(o1.id(), o2.id())
```

Comment:
Can we use `o1.equals(o2)` instead here? At least I did not spot the difference.

Reply:
(I see now that it is copied, but still makes sense I think).

Reply:
Fixed.